### PR TITLE
ZK filter: waiting for full packet in buffer before decoding

### PR DESCRIPTION
**Pull request by Winbobob**
Signed-off-by: Zhewei Hu <zhu@pinterest.com>

Commit Message: ZK filter: waiting for full packet in buffer before decoding
Additional Description: Sometimes ZK filter did not fetch all the data of a certain request or response from buffer at once (since buffer may not contain all the data). Then, ZK filter tried to parse the metadata (len, xid, opcode, etc) from all later parts of the data fetched from the buffer the same way as the first part. Since all later parts of data does not include the metadata, only pure raw data, the decoder error will be recorded. After the discussion, we think it is easier to wait for the all the data for a certain request or response in buffer before decoding.
Risk Level: low
Testing: unit tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
Fixes #Issue: #25024

### CODE REVIEW

**Comment by Winbobob** [comment]
cc @rgs1 and @JuniorHsu, please review, thanks!

**Comment by Winbobob** [comment]
I think another way to implement is to take the advantage of **StopIteration**. If the data in the buffer is not enough, the filter just return StopIteration without drain the buffer and start from the first filter, until the data in the buffer is enough.

In this case, we do not need to introduce a new filter status and do not need the infinite loop.

**Comment by Winbobob** [comment]
cc @fishcakez, just updated the code, thanks!

**Review by JuniorHsu — COMMENTED** [review]
Thanks for moving forward

File: test/extensions/filters/network/zookeeper_proxy/filter_test.cc [hunk 1]
```diff
@@ -1046,6 +1078,33 @@ TEST_F(ZooKeeperFilterTest, MissingXid) {
...
+
+  Buffer::OwnedImpl data =
+      encodeCreateRequestWithPartialData("/foo", "bar", false, enumToSignedInt(OpCodes::Create));
+
+  EXPECT_EQ(Envoy::Network::FilterStatus::StopIteration, filter_->onData(data, false));
+  EXPECT_EQ(0UL, config_->stats().create_rq_.value());
+  EXPECT_EQ(0UL, config_->stats().request_bytes_.value());
+  EXPECT_EQ(0UL, config_->stats().decoder_error_.value());
+}
+
```
[the remark this answers is not in the document]
Comment (Winbobob) [reply test/extensions/filters/network/zookeeper_proxy/filter_test.cc:1092]: Thank you for pointing the 2nd case out!
For the 2nd case, we cannot guarantee that on the 3rd `onData` we have exact the REQ 2 data and not including REQ3 data. If the 3rd `onData` includes REQ3 data and we still return `StopIteration`, which may affect the performance.
I am thinking to maintain several pointers like `start`, `current`, and `end` to represent the current state in the buffer.

[the remark this answers is not in the document]
Comment (Winbobob) [reply test/extensions/filters/network/zookeeper_proxy/filter_test.cc:1092]: Adding pointers make things more complicated. Let's see how current implementation (w/o pointers) goes.

File: source/extensions/filters/network/zookeeper_proxy/decoder.h [hunk 2]
```diff
@@ -136,6 +137,7 @@ class DecoderImpl : public Decoder, Logger::Loggable<Logger::Id::filter> {
     MonotonicTime start_time;
   };
 
+  Network::FilterStatus decodeAfterBuffer(Buffer::Instance& data, DecodeType dtype);
```
Comment (JuniorHsu) [inline source/extensions/filters/network/zookeeper_proxy/decoder.h:140]: add comment here, since `decodeAfterBuffer` is not intuitive.

**Review by JuniorHsu — CHANGES_REQUESTED** [review]

File: source/extensions/filters/network/zookeeper_proxy/decoder.cc [hunk 3]
```diff
@@ -467,9 +467,41 @@ void DecoderImpl::skipStrings(Buffer::Instance& data, uint64_t& offset) {
...
+Network::FilterStatus DecoderImpl::onWrite(Buffer::Instance& data) {
+  return decodeAfterBuffer(data, DecodeType::WRITE);
+}
 
-void DecoderImpl::onWrite(Buffer::Instance& data) { decode(data, DecodeType::WRITE); }
+Network::FilterStatus DecoderImpl::decodeAfterBuffer(Buffer::Instance& data, DecodeType dtype) {
+  uint64_t offset = 0;
+  int32_t len = 0;
+
+  while (offset < data.length()) {
```
[the remark this answers is not in the document]
Comment (Winbobob) [reply source/extensions/filters/network/zookeeper_proxy/decoder.cc:482]: If the buffer contains multiple packets (no matter the last packet's data is entirely in the buffer or not), we will have multiple iterations.
Yes, the decoder are able to handle multiple requests or responses in the buffer.

**Review by JuniorHsu — APPROVED** [review]
Thanks.

**Comment by JuniorHsu** [comment]
@snowp gentle ping as you're the code owner who's able to merge :)

**Review by rgs1 — APPROVED** [review]
Thanks! (sorry for the delay, I am OOO)

**Comment by rgs1** [comment]
One last thing: did you manage to bake this with real ZK traffic?

**Comment by Winbobob** [comment]
> One last thing: did you manage to bake this with real ZK traffic?

Yes, I tested this patch on two clusters.
- One cluster's decoding errors got eliminated with this patch. And there is no latency spike.
- Another cluster still had very few decoding errors with the patch (will look into it later). The latency raised a bit, not sure how the "stop and buffer" mechanism affected the latency. I think the latency measures the time ZK service processing the request, should not be directly related to the filter's behavior? Since the filter tries send the entire packet to ZK service.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**